File: src/fsharp/examples/nbody/NBody.Common.fs
(**
Functionality common to the different implementations.
*)
[<AutoOpen>]
module Tutorial.Fs.examples.NBody.Common

open Alea.CUDA
open FsUnit

(**
Abstract simulator type.
*)
type ISimulator =
    abstract Description : string
    abstract Integrate : newPos:deviceptr<float4> -> oldPos:deviceptr<float4> -> vel:deviceptr<float4> -> numBodies:int -> deltaTime:float32 -> softeningSquared:float32 -> damping:float32 -> unit

(**
Abstract simulation testing type.
*)
type ISimulatorTester =
    abstract Description : string
    abstract Integrate : pos:float4[] -> vel:float4[] -> numBodies:int -> deltaTime:float32 -> softeningSquared:float32 -> damping:float32 -> steps:int -> unit

let float4Zero = float4(0.0f, 0.0f, 0.0f, 0.0f)

(**
Initialize particles randomly in a cube with random velocities but zero momentum.
*)
let initializeBodies1 clusterScale velocityScale numBodies = 
    let rng = System.Random(42)
    let random a b = rng.NextDouble()*a - b |> float32
    let scale = clusterScale * max 1.0f (float32 numBodies/1024.0f)
    let vscale = velocityScale*scale

    let randomPos _ = scale * random 1.0 0.5
    let randomVel _ = vscale * random 1.0 0.5

    let pos = Array.init numBodies (fun _ -> float4(randomPos(), randomPos(), randomPos() + 50.0f, 1.0f))
    let vel = Array.init numBodies (fun _ -> float4(randomVel(), randomVel(), randomVel(), 1.0f))
    let totalMomentum = vel |> Array.fold (fun (acc:float4) e -> 
        float4(acc.x + e.x*e.w, acc.y + e.y*e.w, acc.z + e.z*e.w, acc.w + e.w)) float4Zero

    pos, vel |> Array.map (fun e ->
        float4(e.x - totalMomentum.x/totalMomentum.w/e.w,
               e.y - totalMomentum.y/totalMomentum.w/e.w,
               e.z - totalMomentum.z/totalMomentum.w/e.w,
               e.w)) // initialize with total momentum = 0

(**
Initialize particles randomly in two cubes with angular momentum, s.t. some galaxy similar patterns emerge. Total momentum is set to zero.
*)
let initializeBodies2 clusterScale velocityScale numBodies = 
    let rng = System.Random(42)
    let random a b = rng.NextDouble()*a - b |> float32
    let scale = clusterScale * max 1.0f (float32 numBodies/1024.0f)
    let vscale = velocityScale*scale

    let randomPos _ = scale * random 1.0 0.5
    let randomVel _ = vscale * random 1.0 0.5

    let pos = Array.init numBodies (fun i -> 
        if i < numBodies/2 then float4(randomPos() + 0.5f*scale, randomPos(), randomPos() + 50.0f, 1.0f)
        else float4(randomPos() - 0.5f*scale, randomPos(), randomPos() + 50.0f, 1.0f) )
    let vel = Array.init numBodies (fun i -> 
        if i < numBodies/2 then float4(randomVel(), randomVel() + 0.01f*vscale*pos.[i].x*pos.[i].x, randomVel(), 1.0f)
        else float4(randomVel(), randomVel() - 0.01f*vscale*pos.[i].x*pos.[i].x, randomVel(), 1.0f) )
    let totalMomentum = vel |> Array.fold (fun (acc:float4) e -> 
        float4(acc.x + e.x*e.w, acc.y + e.y*e.w, acc.z + e.z*e.w, acc.w + e.w)) float4Zero

    pos, vel |> Array.map (fun e -> 
        float4(e.x - totalMomentum.x/totalMomentum.w/e.w, 
               e.y - totalMomentum.y/totalMomentum.w/e.w, 
               e.z - totalMomentum.z/totalMomentum.w/e.w, 
               e.w)) // initialize with total momentum = 0

(**
Initialize particles randomly in two cubes with random velocities angular momentum, s.t. some galaxy similar patterns emerge.
Give different particles random mass. Do not set a seed for the random number generator, hence simulations will have different behavour each call. Total momentum is set to zero.
*)
let initializeBodies3 clusterScale velocityScale numBodies = 
    let rng = System.Random()
    let random a b = rng.NextDouble()*a - b |> float32
    let scale = clusterScale * max 1.0f (float32 numBodies/1024.0f)
    let vscale = velocityScale*scale

    let randomPos _ = scale * random 1.0 0.5
    let randomVel _ = vscale * random 1.0 0.5
    let randomMass _ =
        let maxv = 1.3
        let minv = 0.7
        (rng.NextDouble() * (maxv - minv) + minv) |> float32

    let pos = Array.init numBodies (fun i -> 
        if i < numBodies/2 then float4(randomPos() + 0.5f*scale, randomPos(), randomPos() + 50.0f, randomMass())
        else float4(randomPos() - 0.5f*scale, randomPos(), randomPos() + 50.0f, randomMass()) )
    let vel = Array.init numBodies (fun i -> 
        if i < numBodies/2 then float4(randomVel(), randomVel() + 0.01f*vscale*pos.[i].x*pos.[i].x, randomVel(), 1.0f)
        else float4(randomVel(), randomVel() - 0.01f*vscale*pos.[i].x*pos.[i].x, randomVel(), 1.0f) )
    let totalMomentum = vel |> Array.fold (fun (acc:float4) e -> 
        float4(acc.x + e.x*e.w, acc.y + e.y*e.w, acc.z + e.z*e.w, acc.w + e.w)) float4Zero

    pos, vel |> Array.map (fun e -> 
        float4(e.w*e.x - totalMomentum.x/totalMomentum.w/e.w,
               e.w*e.y - totalMomentum.y/totalMomentum.w/e.w, 
               e.w*e.z - totalMomentum.z/totalMomentum.w/e.w, 
               e.w)) // initialize with total momentum = 0

(**
Testing if two simulations behave the same up to some tolerance.
*)
let test (expectedSimulator:ISimulatorTester) (actualSimulator:ISimulatorTester) numBodies =
    let clusterScale = 1.0f
    let velocityScale = 1.0f
    let deltaTime = 0.001f
    let softeningSquared = 0.00125f
    let damping = 0.9995f
    let steps = 5

    printfn "Testing %A against %A with %d bodies..." actualSimulator.Description expectedSimulator.Description numBodies

    let verify (expected:float4[]) (actual:float4[]) (tol:float) =
        (expected, actual) ||> Array.iter2 (fun expected actual ->
            actual.x |> should (equalWithin tol) expected.x
            actual.y |> should (equalWithin tol) expected.y
            actual.z |> should (equalWithin tol) expected.z
            actual.w |> should (equalWithin tol) expected.w )

    let test (tol:float) (initializeBodies:int -> float4[] * float4[]) =
        let expectedPos, expectedVel = initializeBodies numBodies
        let actualPos = Array.copy expectedPos
        let actualVel = Array.copy expectedVel
        for i = 1 to steps do
            expectedSimulator.Integrate expectedPos expectedVel numBodies deltaTime softeningSquared damping 1
            actualSimulator.Integrate actualPos actualVel numBodies deltaTime softeningSquared damping 1
            verify expectedPos actualPos tol
            verify expectedVel actualVel tol
    
    initializeBodies1 clusterScale velocityScale |> test 1e-5
    initializeBodies3 clusterScale velocityScale |> test 1e-4

(**
Measure the performence of a simulation method.
*)
let performance (simulator:ISimulatorTester) numBodies =
    let clusterScale = 1.0f
    let velocityScale = 1.0f
    let deltaTime = 0.001f
    let softeningSquared = 0.00125f
    let damping = 0.9995f
    let steps = 10

    printfn "Perfomancing %A with %d bodies..." simulator.Description numBodies

    let test (initializeBodies:int -> float4[] * float4[]) =
        let pos, vel = initializeBodies numBodies
        simulator.Integrate pos vel numBodies deltaTime softeningSquared damping steps
    
    initializeBodies1 clusterScale velocityScale |> test

(**
Calculating the acceleration $a_{ij}$ particle at position `bj` exerts on particle with position `bi`, and adds it to the acceleration `ai` (resulting in total acceleration on particle with position `bi`, after calling this method for all particles `bj`) which will be returned.
It is used for the CPU as well as for both GPU implementations.

The acceleration is calculated as:

$$$
\begin{equation}
    a_{ij} = \frac{f_{ij}}{m_i} = \frac{r_{ij} m_j}{\sqrt{\left||r_{ij}\right||^2 + \varepsilon^2}^3},
\end{equation}

note, we use artificial units where $G$ is set to 1.

CUDA datatypes `float3` (acceleration) and `float4` (position, 4-th element is the particles mass) are used. 
The function `__nv_rsqrtf` calculates $f(x) = \frac{1}{\sqrt{x}}$ faster and with less registers than calling the two functions seperately.
*)
[<ReflectedDefinition>]
let bodyBodyInteraction softeningSquared (ai:float3) (bi:float4) (bj:float4) =
    // r_ij  [3 FLOPS]
    let r = float3(bj.x - bi.x, bj.y - bi.y, bj.z - bi.z)

    // distSqr = dot(r_ij, r_ij) + EPS^2  [6 FLOPS]
    let distSqr = r.x*r.x + r.y*r.y + r.z*r.z + softeningSquared

    // invDistCube =1/distSqr^(3/2)  [4 FLOPS (2 mul, 1 sqrt, 1 inv)]
    let invDist = __nv_rsqrtf distSqr
    let invDistCube =  invDist * invDist * invDist
    
    // s = m_j * invDistCube [1 FLOP]
    let s = bj.w * invDistCube

    // a_i =  a_i + s * r_ij [6 FLOPS]
    float3(ai.x + r.x*s, ai.y + r.y*s, ai.z + r.z*s)

File: src/csharp/examples/nbody/NBody.Common.cs
//[startCommon]
using System;
using System.Linq;
using Alea.CUDA;
using NUnit.Framework;

namespace Tutorial.Cs.examples.nbody
{
    //[/startCommon]

    //[ISimulator]
    public interface ISimulator
    {
        string Description { get; }
        void Integrate(deviceptr<float4> newPos, deviceptr<float4> oldPos, deviceptr<float4> vel, int numBodies, float deltaTime, float softeningSquared, float damping);
    }
    //[/ISimulator]

    //[ISimulatorTester]
    public interface ISimulatorTester
    {
        string Description { get; }
        void Integrate(float4[] pos, float4[] vel, int numBodies, float deltaTime, float softeningSquared, float damping, int steps);
    }
    //[/ISimulatorTester]

    public abstract class BodyInitializer
    {
        abstract public int NumBodies { get; set; }
        abstract public float PScale { get; set; }
        abstract public float VScale { get; set; }
        public abstract float4 Position(int i);
        public abstract float4 Velocity(float4 position, int i);

        static float4 Momentum(float4 velocity)
        {
            // we store mass in velocity.w
            var mass = velocity.w;
            return new float4(velocity.x * mass,
                              velocity.y * mass,
                              velocity.z * mass,
                              mass);
        }

        //[adjustMomentum]
        static public void Initialize(BodyInitializer initializer, float clusterScale, float velocityScale, int numBodies,
            out float4[] positions, out float4[] velocities)
        {
            var pscale = clusterScale*Math.Max(1.0f, numBodies/1024.0f);
            var vscale = velocityScale*pscale;
            initializer.NumBodies = numBodies;
            initializer.PScale = pscale;
            initializer.VScale = vscale;
            positions = Enumerable.Range(0, numBodies).Select(initializer.Position).ToArray();
            velocities = positions.Select(initializer.Velocity).ToArray();

            // now we try to adjust velocity to make total momentum = zero.
            var momentums = velocities.Select(Momentum).ToArray();
            var totalMomentum = momentums.Aggregate(new float4(0.0f, 0.0f, 0.0f, 0.0f),
                (accum, momentum) =>
                    new float4(accum.x + momentum.x,
                               accum.y + momentum.y,
                               accum.z + momentum.z, 
                               accum.w + momentum.w));
            Console.WriteLine("total momentum and mass 0 = {0}", totalMomentum);

            // adjust velocities
            velocities = velocities.Select((vel, i) => new float4(
                vel.x - totalMomentum.x / totalMomentum.w,
                vel.y - totalMomentum.y / totalMomentum.w,
                vel.z - totalMomentum.z / totalMomentum.w,
                vel.w)).ToArray();

            // see total momentum after adjustment
            momentums = velocities.Select(Momentum).ToArray();
            totalMomentum = momentums.Aggregate(new float4(0.0f, 0.0f, 0.0f, 0.0f),
                (accum, momentum) =>
                    new float4(accum.x + momentum.x,
                               accum.y + momentum.y,
                               accum.z + momentum.z,
                               accum.w + momentum.w));
            Console.WriteLine("total momentum and mass 1 = {0}", totalMomentum);
        }
        //[/adjustMomentum]
    }

    //[initializeBodies1]
    public class BodyInitializer1 : BodyInitializer
    {
        private readonly Random _random = new Random(42);

        public override int NumBodies { get; set; }
        public override float PScale { get; set; }
        public override float VScale { get; set; }

        private float Rand(float scale, float location)
        {
            return (float) (_random.NextDouble()*scale + location);
        }

        private float RandP()
        {
            return PScale*Rand(1.0f, -0.5f);
        }

        private float RandV()
        {
            return VScale*Rand(1.0f, -0.5f);
        }

        public override float4 Position(int i)
        {
            return new float4(RandP(), RandP(), RandP() + 50.0f, 1.0f);
        }

        public override float4 Velocity(float4 position, int i)
        {
            return new float4(RandV(), RandV(), RandV(), 1.0f);
        }
    }
    //[/initializeBodies1]

    //[initializeBodies2]
    public class BodyInitializer2 : BodyInitializer
    {
        private readonly Random _random = new Random(42);

        public override int NumBodies { get; set; }
        public override float PScale { get; set; }
        public override float VScale { get; set; }

        private float Rand(float scale, float location)
        {
            return (float)(_random.NextDouble() * scale + location);
        }

        private float RandP()
        {
            return PScale * Rand(1.0f, -0.5f);
        }

        private float RandV()
        {
            return VScale * Rand(1.0f, -0.5f);
        }

        public override float4 Position(int i)
        {
            if (i < NumBodies / 2)
            {
                return new float4(RandP() + 0.5f * PScale, RandP(), RandP() + 50.0f, 1.0f);
            }
            else
            {
                return new float4(RandP() - 0.5f * PScale, RandP(), RandP() + 50.0f, 1.0f);
            }
        }

        public override float4 Velocity(float4 position, int i)
        {
            if (i < NumBodies / 2)
            {
                return new float4(RandV(), RandV() + 0.01f * VScale * position.x * position.x, RandV(), 1.0f);
            }
            else
            {
                return new float4(RandV(), RandV() - 0.01f * VScale * position.x * position.x, RandV(), 1.0f);
            }
        }
    }
    //[/initializeBodies2]

    //[initializeBodies3]
    public class BodyInitializer3 : BodyInitializer
    {
        private readonly Random _random = new Random();

        public override int NumBodies { get; set; }
        public override float PScale { get; set; }
        public override float VScale { get; set; }

        private float Rand(float scale, float location)
        {
            return (float)(_random.NextDouble() * scale + location);
        }

        private float RandP()
        {
            return PScale * Rand(1.0f, -0.5f);
        }

        private float RandV()
        {
            return VScale * Rand(1.0f, -0.5f);
        }

        private float RandM()
        {
            return Rand(0.6f, 0.7f);
        }

        public override float4 Position(int i)
        {
            if (i < NumBodies / 2)
            {
                return new float4(RandP() + 0.5f * PScale, RandP(), RandP() + 50.0f, RandM());
            }
            else
            {
                return new float4(RandP() - 0.5f * PScale, RandP(), RandP() + 50.0f, RandM());
            }
        }

        public override float4 Velocity(float4 position, int i)
        {
            if (i < NumBodies / 2)
            {
                return new float4(RandV(), RandV() + 0.01f * VScale * position.x * position.x, RandV(), position.w);
            }
            else
            {
                return new float4(RandV(), RandV() - 0.01f * VScale * position.x * position.x, RandV(), position.w);
            }
        }
    }
    //[/initializeBodies3]

    public static class Common
    {
        //[commonTester]
        public static void Test(ISimulatorTester expectedSimulator, ISimulatorTester actualSimulator, int numBodies)
        {
            const float clusterScale = 1.0f;
            const float velocityScale = 1.0f;
            const float deltaTime = 0.001f;
            const float softeningSquared = 0.00125f;
            const float damping = 0.9995f;
            const int steps = 5;

            Console.WriteLine("Testing {0} against {1} with {2} bodies...", actualSimulator.Description,
                expectedSimulator.Description, numBodies);

            {
                float4[] expectedPos, expectedVel;
                BodyInitializer.Initialize(new BodyInitializer1(), clusterScale, velocityScale, numBodies, out expectedPos, out expectedVel);

                for (var i = 0; i < steps; i++)
                {
                    const double tol = 1e-5;
                    var actualPos = new float4[numBodies];
                    var actualVel = new float4[numBodies];
                    Array.Copy(expectedPos, actualPos, numBodies);
                    Array.Copy(expectedVel, actualVel, numBodies);
                    expectedSimulator.Integrate(expectedPos, expectedVel, numBodies, deltaTime, softeningSquared, damping, 1);
                    actualSimulator.Integrate(actualPos, actualVel, numBodies, deltaTime, softeningSquared, damping, 1);
                    for (var j = 0; j < expectedPos.Length; j++)
                    {
                        Assert.AreEqual(actualPos[j].x, expectedPos[j].x, tol);
                        Assert.AreEqual(actualPos[j].y, expectedPos[j].y, tol);
                        Assert.AreEqual(actualPos[j].z, expectedPos[j].z, tol);
                        Assert.AreEqual(actualPos[j].w, expectedPos[j].w, tol);
                    }
                }
            }

            {
                float4[] expectedPos, expectedVel;
                BodyInitializer.Initialize(new BodyInitializer3(), clusterScale, velocityScale, numBodies,
                    out expectedPos, out expectedVel);

                for (var i = 0; i < steps; i++)
                {
                    const double tol = 1e-4;
                    var actualPos = new float4[numBodies];
                    var actualVel = new float4[numBodies];
                    Array.Copy(expectedPos, actualPos, numBodies);
                    Array.Copy(expectedVel, actualVel, numBodies);
                    expectedSimulator.Integrate(expectedPos, expectedVel, numBodies, deltaTime, softeningSquared,
                        damping, 1);
                    actualSimulator.Integrate(actualPos, actualVel, numBodies, deltaTime, softeningSquared, damping, 1);
                    for (var j = 0; j < expectedPos.Length; j++)
                    {
                        Assert.AreEqual(actualPos[j].x, expectedPos[j].x, tol);
                        Assert.AreEqual(actualPos[j].y, expectedPos[j].y, tol);
                        Assert.AreEqual(actualPos[j].z, expectedPos[j].z, tol);
                        Assert.AreEqual(actualPos[j].w, expectedPos[j].w, tol);
                    }
                }
            }
        }
        //[/commonTester]

        //[commonPerfTester]
        public static void Performance(ISimulatorTester simulator, int numBodies)
        {
            const float clusterScale = 1.0f;
            const float velocityScale = 1.0f;
            const float deltaTime = 0.001f;
            const float softeningSquared = 0.00125f;
            const float damping = 0.9995f;
            const int steps = 10;

            Console.WriteLine("Perfomancing {0} with {1} bodies...", simulator.Description, numBodies);

            float4[] pos, vel;
            BodyInitializer.Initialize(new BodyInitializer1(), clusterScale, velocityScale, numBodies, out pos, out vel);
            simulator.Integrate(pos, vel, numBodies, deltaTime, softeningSquared, damping, steps);
        }
        //[/commonPerfTester]


        //[bodyBodyInteraction]
        public static float3 BodyBodyInteraction(float softeningSquared, float3 ai, float4 bi, float4 bj) 
        {
            // r_ij  [3 FLOPS]
            var r = new float3(bj.x - bi.x, bj.y - bi.y, bj.z - bi.z);

            // distSqr = dot(r_ij, r_ij) + EPS^2  [6 FLOPS]
            var distSqr = r.x * r.x + r.y * r.y + r.z * r.z + softeningSquared;

            // invDistCube =1/distSqr^(3/2)  [4 FLOPS (2 mul, 1 sqrt, 1 inv)]
            var invDist = LibDevice.__nv_rsqrtf(distSqr);
            var invDistCube = invDist * invDist * invDist;
    
            // s = m_j * invDistCube [1 FLOP]
            var s = bj.w * invDistCube;

            // a_i =  a_i + s * r_ij [6 FLOPS]
            return(new float3(ai.x + r.x*s, ai.y + r.y*s, ai.z + r.z*s));
        }
        //[/bodyBodyInteraction]
    }
}